### ISSUE TITLE
"indices not found"-style errors when detectors were created via API

### ISSUE THREAD

**Issue by mmguero** [opensearch-project/anomaly-detection-dashboards-plugin#156]
**Describe the bug**

With the v1.2.0 release I was watching closely the fix for #109 (some more details in #155  which I logged and then closed as a duplicate). I am still unable to view the anomaly detection dashboard after creating detectors using the  [create detector API](https://opensearch.org/docs/latest/monitoring-plugins/ad/api/#create-anomaly-detector). The dashboard UI shows `Unable to get all detectors` errors, and the console output throws various exceptions about indices not being found (`opensearch-ad-plugin-result-*` and/or `.opendistro-alerting-config`).

**To Reproduce**

1. Create a detector using the [create detector API](https://opensearch.org/docs/latest/monitoring-plugins/ad/api/#create-anomaly-detector) via `curl` or some similar tool, not via the plugin UI
2. Validate that the detector was created correctly from the API output and with the [search detector API](https://opensearch.org/docs/latest/monitoring-plugins/ad/api/#search-detector)
3. Open Dashboards and navigate to the anomaly detection dashboards plugin UI
4. The `Unable to get all detectors` errors are displayed in the UI and the following exceptions are written to the console output dashboards:

**Anomaly Detection Dashboard UI**
```
Anomaly detector - Unable to search detectors { Error: [illegal_argument_exception] No indices found
    at respond (/usr/share/opensearch-dashboards/node_modules/elasticsearch/src/lib/transport.js:349:15)
    at checkRespForFailure (/usr/share/opensearch-dashboards/node_modules/elasticsearch/src/lib/transport.js:306:7)
    at HttpConnector.<anonymous> (/usr/share/opensearch-dashboards/node_modules/elasticsearch/src/lib/connectors/http.js:173:7)
    at IncomingMessage.wrapper (/usr/share/opensearch-dashboards/node_modules/lodash/lodash.js:4991:19)
    at IncomingMessage.emit (events.js:203:15)
    at endReadableNT (_stream_readable.js:1145:12)
    at process._tickCallback (internal/process/next_tick.js:63:19)
  status: 400,
  displayName: 'BadRequest',
  message: '[illegal_argument_exception] No indices found',
  path:
   '/_plugins/_anomaly_detection/detectors/results/_search/opensearch-ad-plugin-result-*?only_query_custom_result_index=false',
  query: {},
  body:
   { error:
      { root_cause: [Array],
        type: 'illegal_argument_exception',
        reason: 'No indices found' },
     status: 400 },
  statusCode: 400,
  response:
   '{"error":{"root_cause":[{"type":"illegal_argument_exception","reason":"No indices found"}],"type":"illegal_argument_exception","reason":"No indices found"},"status":400}',
  toString: [Function],
  toJSON: [Function] }
```

**Anomaly Detection Detectors UI**
```
[2021-11-29T17:31:08,810][ERROR][o.o.a.u.AlertingException] [opensearch] Alerting error: [.opendistro-alerting-config] IndexNotFoundException[no such index [.opendistro-alerting-config]]
Unable to get monitor on top of detector { Error: [alerting_exception] Configured indices are not found: [.opendistro-alerting-config]
    at respond (/usr/share/opensearch-dashboards/node_modules/elasticsearch/src/lib/transport.js:349:15)
    at checkRespForFailure (/usr/share/opensearch-dashboards/node_modules/elasticsearch/src/lib/transport.js:306:7)
    at HttpConnector.<anonymous> (/usr/share/opensearch-dashboards/node_modules/elasticsearch/src/lib/connectors/http.js:173:7)
    at IncomingMessage.wrapper (/usr/share/opensearch-dashboards/node_modules/lodash/lodash.js:4991:19)
    at IncomingMessage.emit (events.js:203:15)
    at endReadableNT (_stream_readable.js:1145:12)
    at process._tickCallback (internal/process/next_tick.js:63:19)
  status: 404,
  displayName: 'NotFound',
  message:
   '[alerting_exception] Configured indices are not found: [.opendistro-alerting-config]',
  path: '/_plugins/_alerting/monitors/_search',
  query: {},
  body:
   { error:
      { root_cause: [Array],
        type: 'alerting_exception',
        reason:
         'Configured indices are not found: [.opendistro-alerting-config]',
        caused_by: [Object] },
     status: 404 },
  statusCode: 404,
  response:
   '{"error":{"root_cause":[{"type":"alerting_exception","reason":"Configured indices are not found: [.opendistro-alerting-config]"}],"type":"alerting_exception","reason":"Configured indices are not found: [.opendistro-alerting-config]","caused_by":{"type":"exception","reason":"org.opensearch.index.IndexNotFoundException: no such index [.opendistro-alerting-config]"}},"status":404}',
  toString: [Function],
  toJSON: [Function] }
Anomaly detector - Unable to search detectors { Error: [illegal_argument_exception] No indices found
    at respond (/usr/share/opensearch-dashboards/node_modules/elasticsearch/src/lib/transport.js:349:15)
    at checkRespForFailure (/usr/share/opensearch-dashboards/node_modules/elasticsearch/src/lib/transport.js:306:7)
    at HttpConnector.<anonymous> (/usr/share/opensearch-dashboards/node_modules/elasticsearch/src/lib/connectors/http.js:173:7)
    at IncomingMessage.wrapper (/usr/share/opensearch-dashboards/node_modules/lodash/lodash.js:4991:19)
    at IncomingMessage.emit (events.js:203:15)
    at endReadableNT (_stream_readable.js:1145:12)
    at process._tickCallback (internal/process/next_tick.js:63:19)
  status: 400,
  displayName: 'BadRequest',
  message: '[illegal_argument_exception] No indices found',
  path:
   '/_plugins/_anomaly_detection/detectors/results/_search/opensearch-ad-plugin-result-*?only_query_custom_result_index=false',
  query: {},
  body:
   { error:
      { root_cause: [Array],
        type: 'illegal_argument_exception',
        reason: 'No indices found' },
     status: 400 },
  statusCode: 400,
  response:
   '{"error":{"root_cause":[{"type":"illegal_argument_exception","reason":"No indices found"}],"type":"illegal_argument_exception","reason":"No indices found"},"status":400}',
  toString: [Function],
  toJSON: [Function] }
```

**Expected behavior**
Detectors that were successfully created and exist should be displayed in the UI. Either the  [create detector API](https://opensearch.org/docs/latest/monitoring-plugins/ad/api/#create-anomaly-detector) should create the necessary indices (in which case this bug should be moved to that repo) or the anomaly detection dashboards plugin should handle it gracefully. 

**Plugins**
Please list all plugins currently enabled.

For dashboards:
```
$ /usr/share/opensearch-dashboards/bin/opensearch-dashboards-plugin list
alertingDashboards@1.2.0.0
anomalyDetectionDashboards@1.2.0.0
ganttChartDashboards@1.2.0.0
indexManagementDashboards@1.2.0.0
kbnSankeyVis@1.2.0.0
observabilityDashboards@1.2.0.0
queryWorkbenchDashboards@1.2.0.0
reportsDashboards@1.2.0.0
```

For opensearch:
```
$ /usr/share/opensearch/bin/opensearch-plugin list
opensearch-alerting
opensearch-anomaly-detection
opensearch-asynchronous-search
opensearch-cross-cluster-replication
opensearch-index-management
opensearch-job-scheduler
opensearch-knn
opensearch-observability
opensearch-performance-analyzer
opensearch-reports-scheduler
opensearch-sql

```

**Host/Environment (please complete the following information):**

- OS: Docker based on official `opensearchproject/opensearch:1.2.0` and `opensearchproject/opensearch-dashboards:1.2.0` images, running on top of Debian 11 x86_64.
- Version 1.2.0

**Additional Context:**

This is actually worse than what I was experiencing in v1.1.0. For that, at least you had a workaround where starting/stopping a detector job would cause the necessary indices to be created. I have no workaround at the moment.

**Comment by ylwu-amzn** [comment]
hi, @mmguero, thanks for reporting this issue. The detector list page will search both detectors and anomaly results. When no detector started, then no AD result indices created on new cluster. So the backend code will throw "No indices found" exception when search anomaly results.

One workaround is you can create a test cluster with custom result index. A custom result index will be created when create detector. Then the search AD result API won't throw error and detector list can show up. This is an example of detector using custom result index:
```
POST _plugins/_anomaly_detection/detectors
{
  "name": "test",
  "description": "Test detector",
  "time_field": "timestamp",
  "indices": [
    "server_log"
  ],
  "filter_query": {
    "bool": {
      "filter": [
        {
          "range": {
            "value": {
              "gte": 1
            }
          }
        }
      ]
    }
  },
  "detection_interval": {
    "period": {
      "interval": 1,
      "unit": "Minutes"
    }
  },
  "window_delay": {
    "period": {
      "interval": 1,
      "unit": "Minutes"
    }
  },
  "feature_attributes": [
    {
      "feature_name": "test",
      "feature_enabled": true,
      "aggregation_query": {
        "test": {
          "sum": {
            "field": "value"
          }
        }
      }
    }
  ],
  "result_index": "opensearch-ad-plugin-result-test"
}
```

**Comment by mmguero** [comment]
Thanks for the suggestion and update on the fix!

**Comment by ohltyler** [comment]
@mmguero the patch release is now [live](https://github.com/opensearch-project/anomaly-detection/releases/tag/1.2.0.1) - feel free to download the attached zip and manually installing. Please reach out if you see any more issues, and thanks again for reporting!

**Comment by ohltyler** [comment]
Closing since this has been fixed as part of [backend PR #322](https://github.com/opensearch-project/anomaly-detection/pull/322).
